Initial attempt at a package for msgpack-c.

bluescarni opened the pull request.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/msgpack-c```) and found some lint.

Here's what I've got...


For **recipes/msgpack-c**:

 * about/license_family 'Boost' not allowed. Allowed families are AGPL,
APACHE, BSD, GPL, GPL2, GPL3, LGPL, MIT, NONE, OTHER, PROPRIETARY,
PSF, and PUBLICDOMAIN.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/msgpack-c```) and found it was in an excellent condition.

isuruf commented on recipes/msgpack-c/meta.yaml:
```diff
@@ -0,0 +1,38 @@
+test:
+  requires:
+    - cmake
+    - toolchain
```
Packages above are not needed for tests.
For tests can you check for a include file and the library installed?

isuruf commented on recipes/msgpack-c/meta.yaml:
```diff
@@ -0,0 +1,38 @@
+requirements:
+  build:
+    - cmake
+    - toolchain
```
Can you try a build with boost and zlib? I see that they are optional dependencies, but since they are on conda-forge we can use them.

isuruf commented on recipes/msgpack-c/bld.bat:
```diff
@@ -0,0 +1,12 @@
+    -G "%CMAKE_GENERATOR%" ^
+    -DCMAKE_INSTALL_PREFIX=%LIBRARY_PREFIX% ^
+    -DMSGPACK_CXX11=YES ^
+    -DCMAKE_BUILD_TYPE=Release ^
```
appveyor.yml of msgpack-c uses `-DCMAKE_CXX_FLAGS='"/D_VARIADIC_MAX=10 /EHsc"'`

https://github.com/msgpack/msgpack-c/blob/master/appveyor.yml#L49

bluescarni commented:
@isuruf thanks for the feedback!

I have a question: the msgpack build system requires to specify explicitly if one is building for 32bit, otherwise the build will fail. Is there a way to detect 32 vs 64 bit in build.sh/bld.bat?

bluescarni commented:
@isuruf maybe I am wrong about the 32 bit stuff, it seems on appveyor there is some other issue with the last 3 builds. Also, it seems like it's not locating correctly boost in the successful builds (``Could NOT find Boost``) but it seems to work anyway, probably due to the fact that I pass into CMake the correct variables directly?

isuruf commented:
For checking 32-bit or 64-bit you can use `%ARCH%` environment variable which would be 32 or 64.

bluescarni commented:
@isuruf any idea on why the 64bits builds on appveyor are failing?

isuruf commented:
From what I read on the internet, the warning `no object file generated` is misleading and hides another warning. Can you add a patch to remove `/Wx` added in CMakelists.txt and see whether the warnings are harmless?

isuruf commented:
Can you use NMake Makefiles as the generator to fix the Python 2.7 x64 error?

isuruf commented on recipes/msgpack-c/meta.yaml:
```diff
@@ -0,0 +1,46 @@
+    - test -f $PREFIX/lib/libmsgpackc.dylib                     # [osx]
+    - if not exist %LIBRARY_INC%\\msgpack.h exit 1              # [win]
+    - if not exist %LIBRARY_INC%\\msgpack.hpp exit 1            # [win]
+    - if not exist %LIBRARY_LIB%\\msgpackc.lib exit 1           # [win]
```
This is the static library, can you rename this to `msgpackc_static.lib` and the import library `msgpackc_import.lib` to `mspackc.lib`?

bluescarni replied:
Ok.

bluescarni commented:
@isuruf It seems to pass now. There were 2 problems on Win64:

- warnings treated as errors (I suspect consequence of the fact that msgpack's CI does not actually test 64bit windows builds). I fixed this by disabling the specific warnings for the moment;
- the nmake issue on Python 2.7 64bit builds, fixed as you suggested.

The tests also seem to run on unix, but on Windows I could not detect any output from the tests.

isuruf commented on recipes/msgpack-c/meta.yaml:
```diff
@@ -0,0 +1,46 @@
+    - cmake
+    - toolchain
+    - boost
+    - zlib
```
Add `boost` and `zlib` to run requirements. You need to pin `boost` and `zlib`, see https://github.com/conda-forge/piranha-feedstock/blob/master/recipe/meta.yaml#L22-L24

bluescarni replied:
@isuruf what does it mean exactly to "pin"? To specify a specific version of the libraries?

isuruf replied:
Yes, On conda-forge, these dependencies are pinned so that there are no ABI compatibility issues. Once there is a new version, all the feedstocks using these dependencies will be upgraded.

List is below, https://github.com/conda-forge/staged-recipes/wiki/Pinned-dependencies

isuruf replied:
Above was the reason to use that set of specific versions. Pinning is necessary so that an environment with `msgpack-c==2.0.0` would also have `boost==1.62.*`.

bluescarni replied:
You mean that when boost 1.63 is released, all recipes are automatically updated to depend on 1.63? Or should we update them manually?

isuruf replied:
There's a bot running that would create PRs, but they have to be merged manually.

isuruf commented:
You mean the tests checking for the files? On appveyor it is not shown, but it will error if they are not found.

bluescarni commented:
@isuruf yes, I meant the tests checking the files.

isuruf commented on recipes/msgpack-c/meta.yaml:
```diff
@@ -0,0 +1,50 @@
+    - cmake
+    - toolchain
+    - boost 1.62.*
+    - zlib 1.2.*
```
Can you add `python` for `win` here so that this recipe would be built against all the vc versions? You also need to add vc features.
```yaml
build:
  features:
    - vc9   # [win and py27]
    - vc10  # [win and py34]
    - vc14  # [win and py35]

requirements:
  build:
    - python  # [win]
    - vc 9    # [win and py27]
    - vc 10   # [win and py34]
    - vc 14   # [win and py35]
  run:
    - vc 9   # [win and py27]
    - vc 10  # [win and py34]
    - vc 14  # [win and py35]
```

See https://github.com/conda-forge/staged-recipes/wiki/VC-features

bluescarni commented:
@isuruf are 32bit builds considered at all in non-windows platforms?

isuruf commented:
In conda-forge, no. conda has major support for linux-32, but minimal support for osx-32 (https://repo.continuum.io/pkgs/)

bluescarni commented:
@isuruf ok, I was asking because there's a specific switch in the build system to be activated for 32bit builds, but it does something only if the compiler is gcc/clang (in which case, the ``-m32`` flag is set).

bluescarni commented:
@isuruf I have enabled C++11 in the builds, where possible.

isuruf commented:
Looks good to me.

jakirkham commented on recipes/msgpack-c/bld.bat:
```diff
@@ -0,0 +1,36 @@
+    -DCMAKE_BUILD_TYPE=Release ^
+    ..
+
+cmake --build . --config Release
```
Are there tests we can run (e.g. `ctest`)?

isuruf commented on recipes/msgpack-c/meta.yaml:
```diff
@@ -0,0 +1,64 @@
+
+test:
+  requires:
+    - gtest                                                     # [win]
```
This is not needed.

bluescarni commented:
@isuruf @jakirkham There are many segfaults and other errors in appveyor builds marked as "passing". Any idea of what is going on and why the problematic builds are not marked as failing?

isuruf commented:
This might help with unix, https://github.com/msgpack/msgpack-c/commit/61c053d37ea8b2c3e790997205298192db1e6e02

For appveyor, I have no idea why it segfaults. Let's just skip windows for now.

isuruf commented:
@jakirkham, can you have a look?

jakirkham commented:
Went to triggered AppVeyor, but forgot to read this. 😄

> ...windows is skipped anyways.

jakirkham commented on recipes/msgpack-c/meta.yaml:
```diff
@@ -0,0 +1,69 @@
+    # https://github.com/msgpack/msgpack-c/commit/4a1db60af8d6bdb258a34352f5cd8d9800544372.patch
+    - gtest.patch
+    # This is needed for a bug in gcc version used where isnan is ambiguous. (::isnan and std::isnan)
+    - ambiguous_nan_inf.patch   # [linux]
```
Has this gone upstream? Is there a PR or issue open?

isuruf replied:
This bug is not there when using gcc 4.9 or above, but since we are using gcc 4.8 it is there.
This patch doesn't work with osx, so I'll have to find a way to make it work on all platforms before sending it upstream.

isuruf replied:
https://github.com/msgpack/msgpack-c/issues/554

bluescarni replied:
@isuruf the correct way of calling ``isinf()`` and ``isnan()`` in C++ is to include ``<cmath>`` and call ``std::isinf()`` and ``std::isnan()``. I suspect that the patch does not work in OSX because it is using libc++, which is not probably pulling in the ``<cmath>`` header through the inclusion of some other header.

Shall I give it a go? I cannot do it right now but I can update the PR tonight.

isuruf replied:
Sure.

isuruf replied:
I think msgpack-c is including `math.h` because it is testing the C interface in `test/msgpack_c.cpp`. They do it correctly on `test/msgpack_basic.cpp` which tests the C++ interface. I know that including `cmath` will work, but I didn't do it because I thought the difference between `test/msgpack_basic.cpp`  and `test/msgpack_c.cpp`  was intentional, which might not be true.

bluescarni replied:
Still, ``msgpack_c.cpp`` is also including ``<vector>`` which is C++. I think the ambiguity derives from the ``using namespace std;`` statement on top, I am now thinking the better way would probably be to explicitly use ``std::isinf()`` in the code rather than relying on the ``using namespace std;`` statement.

jakirkham commented on recipes/msgpack-c/meta.yaml:
```diff
@@ -0,0 +1,69 @@
+    - gtest
+    - cmake
+    - toolchain
+    - boost 1.62.*
```
We are on `1.63.*` now. So please update this.

jakirkham commented on recipes/msgpack-c/meta.yaml:
```diff
@@ -0,0 +1,69 @@
+    - vc 9   # [win and py27]
+    - vc 10  # [win and py34]
+    - vc 14  # [win and py35]
+    - boost 1.62.*
```
...and this.

jakirkham commented on recipes/msgpack-c/meta.yaml:
```diff
@@ -0,0 +1,69 @@
+  version: {{ version }}
+
+source:
+  fn: cpp-{{ version }}.tar.gz
```
This doesn't have to match the filename in the URL. So it could be `{{ name }}-{{ version }}.tar.gz`, which is handy when you are building locally and trying to figure out what is in your source cache. 😆

jakirkham commented on recipes/msgpack-c/meta.yaml:
```diff
@@ -0,0 +1,69 @@
+    - vc9     # [win and py27]
+    - vc10    # [win and py34]
+    - vc14    # [win and py35]
+  skip: true  # [win]
```
Would change this to `win or py==36`.

isuruf replied:
Since windows is skipped, this is unnecessary.

jakirkham replied:
More thinking of making it easier for one to switch later. Though agreed this could simply be taken as advisement for the future.

bluescarni commented:
@isuruf sorry I am slow at understanding :) I think your approach for the patch is fine, too bad upstream is not much intersted in having it fixed. I just updated the patch for th elatest msgpack-c version and removed the gtest patch (I believe it's been fixed upstream as there are no more build errors).

@jakirkham CI is passing now, I updated boost to 1.63 and bumped up the msgpack-c version to the latest release from upstream.

jakirkham commented on recipes/msgpack-c/meta.yaml:
```diff
@@ -0,0 +1,66 @@
+    - vc 9   # [win and py27]
+    - vc 10  # [win and py34]
+    - vc 14  # [win and py35]
+    - boost 1.63.*
```
So `boost` implies Python dependence. Do we want to use the newly minted `boost-cpp`? 😉

isuruf replied:
I think we should avoid that until `boost` starts using `boost-cpp`, as `msgpack-c` will be used in a C++ library with python bindings and `boost` and `boost-cpp` will start overwriting with each other.

isuruf replied:
Now, boost-cpp can be used. @bluescarni, can you change `boost` to `boost-cpp`? `boost-cpp` is `boost` without the python bindings.

bluescarni replied:
ok, just pushed the change.

jakirkham commented:
This feels like it is in a good place to me. Had a question about using `boost` vs `boost-cpp` (sort of directed @isuruf 😉). Otherwise LGTM.

jakirkham commented:
Thanks @bluescarni.

bluescarni commented:
@jakirkham Cheers!
